### PR TITLE
[WIP] Added jlink/cmsis-dap overviews

### PR DESCRIPTION
**Pull request by ryankurte**
Probably to be split into separate sections when complete? I got sidetracked by other issues before finishing though so it's still WIP.

### CODE REVIEW

File: src/overview.md [hunk 1]
```diff
@@ -66,16 +66,42 @@ And as we talked about JTAG and SWD earlier, OpenOCD supports both of these sign
...
+
+### [ARM DAPLink]
+
+DAPLink is a project by ARM to develop an open source cortex debug probe, this provides a set of interfaces to simplify programming and debugging and can be implemented on nearly any USB-capable hardware. DAPLink provides a set of endpoints including a CMSIS-DAP interface for debugging, a USB disk for drag-and-drop programming, and an optional serial port for communication with the target system.
 
 [DAPLink]:https://os.mbed.com/docs/latest/tools/daplink.html
 
+#### Programming
+
+On connecting a DAPLink device to your system, a USB drive should become available. To flash, copy your binary file (*.bin) to the drive and wait for completion.
```
Comment (therealprof) [inline src/overview.md]: Yeah, that does not work reliably in my experience, so we should not recommend that way. Also in Rust there're no .bin files which could be uploaded.

File: src/overview.md [hunk 2]
```diff
@@ -66,16 +66,42 @@ And as we talked about JTAG and SWD earlier, OpenOCD supports both of these sign
...
+
+#### Debugging
+
+TODO: try it out
```
Comment (therealprof) [inline src/overview.md]: I can take on that part.

File: src/overview.md (hunk 1 above)
Comment (thejpster) [reply src/overview.md]: I use arm-none-eabi-objcopy to produce binaries for Monotron which I can load with lm4flash (or inject into RAM). Maybe it's worth talking about how to do that (shell scripts, cargo-make, cargo issue #545, cargo-binutils, etc).

Comment (therealprof) [reply src/overview.md]: Sounds like a good idea to elaborate on that. I think the problem with the USB MSD emulation is that it cannot cope with irrelevant sections in the produced ELF files and just tries to flash everything.

However I would expect a reliable approach to be able to cope with that and not require additional steps, so I'd put a universal solution (i.e. OpenOCD) that "just works" in the first place.

Comment (ryankurte) [reply src/overview.md]: I've never had any issue flashing `.bin`s via this interface, but, am happy with OpenOCD being the primary method ^_^

> I use arm-none-eabi-objcopy to produce binaries for Monotron ...  Maybe it's worth talking about how to do that

Sounds good, will open another issue.

**Comment by ryankurte** [comment]
If it's alright with y'all I might just merge this as a WIP and we can keep building in more PRs?

File: src/overview.md [hunk 3]
```diff
@@ -108,7 +108,7 @@ Depending on the debugger you use, you may need to convert your compiled output
 
 [cargo-binutils](https://github.com/rust-embedded/cargo-binutils) provides an `objcopy` command to convert between formats, to create a bin file call `cargo objcopy --bin NAME --release -- -O binary NAME.bin` or a hex file call `cargo objcopy --bin NAME --release -- -O hex NAME.hex` where `NAME` is the name of your project. 
 
-You can also use `arm-none-eabi-objcopy` 
+You can also use `arm-none-eabi-objcopy` for the same purpose, with `arm-none-eabi-objcopy -O binary target/.../NAME NAME.bin` to generate binaries and `arm-none-eabi-objcopy -O hex target/.../NAME NAME.hex` to generate hex files. Note that `arm-none-eabi-objcopy` requires the full path to the target file, where `cargo objcopy` can determine the target output directory itself.
```
Comment (therealprof) [inline src/overview.md]: `arm-none-eabi-objcopy -O hex` should be `arm-none-eabi-objcopy -O ihex`

Comment (ryankurte) [reply src/overview.md]: Good catch, thanks!

**Comment by ryankurte** [comment]
For cargo objcopy? Maybe I was trying the wrong set of args but I couldn't get it to work without specifying `--bin` or `--example` pointing to the exact bin or example, I opened an issue  https://github.com/rust-embedded/cargo-binutils/issues/36 about arguments and defaults.

**Comment by therealprof** [comment]
Seems to be a bug in `cargo-binutils`, typically if you have a `src/main.rs` you don't need to specify either `--bin` or `--example`. In fact it's a rather annoying bug that it doesn't work automatically, because for single-binary crates you'll have to specify the crate name itself:
```
# cargo objcopy  --release --bin stm32f4-smoltcp-demo -- -O binary out.bin
   Compiling stm32f4-smoltcp-demo v0.1.0 (/Users/egger/OSS/stm32f4-smoltcp-demo)
    Finished release [optimized + debuginfo] target(s) in 3.42s
```

But what I meant to ask was: Do we have a section somewhere that describes which options to use when to select the correct binary?

**Comment by ryankurte** [comment]
> because for single-binary crates you'll have to specify the crate name itself

yep, that's what i got working / all i intended to convey in those examples

> Do we have a section somewhere that describes which options to use when to select the correct binary?

aha! no we don't, mostly because i couldn't work out what those options should be (and switched back to `arm-none-eabi-objcopy` because of it), it'd be an excellent addition if you're up for it?

**Pull request merged**